impl-tinkerpop/src/main/java/org/hawkular/inventory/impl/tinkerpop/AbstractSourcedGraphService.java

```diff
@@ -104,4 +234,90 @@ protected void removeRelationship(Constants.Type typeInSource, Relationships.Wel
     protected abstract String getProposedId(Blueprint b);
 
     protected abstract Filter[] initNewEntity(Vertex newEntity, Blueprint blueprint);
+
+    private static class BFSTransitiveClosureSpliterator implements Spliterator<Vertex> {
+        private final Direction direction;
+        private final String[] edgeLabels;
+        Deque<Iterator<Vertex>> pending;
+
+        private BFSTransitiveClosureSpliterator(Iterator<Vertex> pending, Direction direction,
+                                                String[] edgeLabels) {
+            this.direction = direction;
+            this.edgeLabels = edgeLabels;
+            this.pending = new ArrayDeque<>(Arrays.asList(pending));
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super Vertex> action) {
+            Iterator<Vertex> queueToProcess = peekNonEmpty();
+            if (queueToProcess == null) {
+                return false;
+            }
+
+            Vertex toProcess = queueToProcess.next();
+
+            pending.add(toProcess.getVertices(direction, edgeLabels).iterator());
+
+            action.accept(toProcess);
```

Comment:
Right, I've just tried on `titan-server-0.4.4` and it removed only v2. Same holds for `v1.out('contains').loop(1){true}{true}.each{g.removeVertex(it)}`

This works for me 
```groovy
v1.out('contains').loop(1){it.loops < 10}{true}.order{-it.a.path.toList().unique{a, b -> a <=> b}.size <=> it.b.path.toList().unique{a, b -> a <=> b}.size}.remove()
v1.remove()
```

It sorts the vertices by the path from v1 (vertices with the longest path go first). So basically it removes everything starting from the leaves.